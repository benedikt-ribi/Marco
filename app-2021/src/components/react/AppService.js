export class AppService {
  constructor(dispatch) {
    this.dispatch = dispatch;
  }

  setMetadata(classMetadata) {
    this.dispatch({ type: "setMetadata", classMetadata });
  }

  setEntity(entity) {
    this.dispatch({ type: "setEntity", entity });
  }

  saveEntity(entity) {
    this.dispatch({ type: "saveEntity", entity });
  }

  deleteEntity(entity) {
    this.dispatch({ type: "deleteEntity", entity });
  }

  createEntity() {
    this.dispatch({ type: "createEntity" });
  }

  changeEntityData(entity) {
    this.dispatch({ type: "changeEntityData", entity });
  }

  setEntities(entities) {
    this.dispatch({ type: "setEntities", entities });
  }
}
