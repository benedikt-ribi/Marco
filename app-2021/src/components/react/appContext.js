import { createContext } from "../../lib/react.development.js";

export const AppContext = createContext(null);
