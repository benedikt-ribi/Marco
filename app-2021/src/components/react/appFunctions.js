import { Crud } from "../../utils/crud.js";
import { createMetadata } from "../../metadataFactory.js";

export function reducer(state, action) {
  switch (action.type) {
    case "setMetadata":
      return resetState(state, action.classMetadata);
    case "setEntity":
      return { ...state, entity: action.entity };
    case "saveEntity": {
      const savedEntity = action.entity;
      const newEntities = state.entities.filter(e => e.id).map(e => (savedEntity.id === e.id ? savedEntity : e));
      if (newEntities.indexOf(savedEntity) === -1) {
        newEntities.push(savedEntity);
      }

      return { ...state, entity: savedEntity, entities: newEntities };
    }
    case "deleteEntity": {
      const entityToDelete = action.entity;
      const entities = state.entities;

      let entityToSelect = state.entity;
      if (entityToDelete.id === state.entity.id) {
        const index = entities.indexOf(entityToDelete);
        if (index < 0) {
          // Do nothing; entityToDelete is not in list
        } else {
          entityToSelect = index > 0 ? entities[index - 1] : entities[index + 1];
        }
      }

      const newEntities = entities.filter(e => e.id !== entityToDelete.id);

      return { ...state, entity: entityToSelect, entities: newEntities };
    }
    case "createEntity": {
      const newEntity = state.classMetadata.createNew();
      return { ...state, entity: newEntity, entities: [...state.entities, newEntity] };
    }
    case "changeEntityData": {
      const { entity } = state;
      const { name, value } = action;
      const newEntity = {
        ...entity,
        [name]: value,
      };

      return { ...state, entity: newEntity };
    }
    case "setEntities":
      return {
        ...state,
        entities: action.entities,
        entity: action.entities.length > 0 ? action.entities[0] : null,
      };
    default:
      throw new Error(`Unknown action [${action.type}]`);
  }
}

export const createInitialState = () => {
  const rootUrl = "http://localhost:3000/";
  const allMetadata = createMetadata(rootUrl);
  const classMetadata = allMetadata[0];

  return resetState({ allMetadata }, classMetadata);
};

function resetState(state, classMetadata) {
  return {
    ...state,
    classMetadata,
    crud: new Crud(classMetadata.rootUrl),
  };
}
