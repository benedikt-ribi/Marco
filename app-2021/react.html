<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <title>React Demo</title>
    <link rel="stylesheet" href="reset.css" />
    <link rel="stylesheet" href="styles.css" />
    <script src="./src/lib/react.development.js"></script>
    <script src="./src/lib/react-dom.development.js"></script>
    <script src="./src/lib/babel.min.js"></script>
    <script type="text/babel" data-type="module">
      import { AppService } from "./src/components/react/AppService.js";
      import { createInitialState, reducer } from "./src/components/react/appFunctions.js";

      const AppContext = React.createContext(null);

      function AppHeader() {
        const { state, service } = React.useContext(AppContext);
        const { allMetadata, classMetadata } = state;

        return (
          <>
            <h1>{classMetadata.pluralCaption}</h1>
            <ul>
              {allMetadata.map(m => (
                <li key={m.caption}>
                  <a href="#" onClick={() => service.setMetadata(m)}>
                    {m.pluralCaption}
                  </a>
                </li>
              ))}
            </ul>
          </>
        );
      }

      function Item({ entity }) {
        const { state, service } = React.useContext(AppContext);
        const { classMetadata, entity: selectedEntity } = state;

        return (
          <>
            <a
              className={entity.id === selectedEntity?.id ? "selected" : ""}
              href="#"
              onClick={() => service.setEntity(entity)}
            >
              {classMetadata.getTitle(entity)}
            </a>
          </>
        );
      }

      function List() {
        const { state } = React.useContext(AppContext);
        const { entities } = state;

        return entities.map(e => <Item key={e.id ?? 0} entity={e} />);
      }

      function InputRow({ p, saveButton }) {
        const { state, service } = React.useContext(AppContext);
        const { entity } = state;

        const inputChange = e => {
          service.changeEntityData(e.target.id, e.target.value);

          saveButton.current.disabled = !e.target.form.checkValidity();
        };

        return (
          <>
            <label htmlFor={p.name}>{p.caption}</label>
            <input
              type={p.type}
              id={p.name}
              required
              placeholder={p.caption}
              value={entity[p.name]}
              onInput={inputChange}
            />
          </>
        );
      }

      function Detail() {
        const { state, service } = React.useContext(AppContext);
        const { classMetadata, entity } = state;
        const saveButton = React.useRef(null);

        const saveEntity = async e => {
          await service.saveEntity(state);
          e.preventDefault();
        };

        const deleteEntity = async e => {
          await service.deleteEntity(state);
          e.preventDefault();
        };

        return !entity ? (
          <div>Nothing selected</div>
        ) : (
          <form>
            {classMetadata.properties.map(p => (
              <InputRow key={p.name} p={p} saveButton={saveButton} />
            ))}
            <span>&nbsp;</span>
            <span>
              <button ref={saveButton} onClick={saveEntity}>
                Save
              </button>
              <button disabled={!entity.id} onClick={deleteEntity}>
                Delete
              </button>
            </span>
          </form>
        );
      }

      function MasterDetail() {
        const { state, service } = React.useContext(AppContext);
        const { classMetadata } = state;
        const [loading, setLoading] = React.useState(true);

        React.useEffect(async () => {
          setLoading(true);
          const data = await state.crud.getAll();
          service.setEntities(data);
          setLoading(false);
        }, [classMetadata]);

        return loading ? (
          <div>loading...</div>
        ) : (
          <>
            <nav>
              <button onClick={() => service.createEntity()}>New...</button>
              <List />
            </nav>
            <article>
              <Detail />
            </article>
          </>
        );
      }

      function App() {
        const [state, dispatch] = React.useReducer(reducer, createInitialState());
        const service = new AppService(dispatch);

        return (
          <AppContext.Provider value={{ state, service }}>
            <header>
              <AppHeader />
            </header>
            <main>
              <MasterDetail />
            </main>
          </AppContext.Provider>
        );
      }

      ReactDOM.render(<App />, document.getElementById("root"));
    </script>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>
