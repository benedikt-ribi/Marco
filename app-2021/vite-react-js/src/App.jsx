import { AppContext } from "./AppContext";
import { AppService } from "../../common/AppService.js";
import { createInitialState, reducer } from "../../common/appFunctions.js";
import { useReducer } from "react";
import { AppHeader } from "./AppHeader";
import { MasterDetail } from "./MasterDetail";
import "../reset.css";
import "../styles.css";

function App() {
  const [state, dispatch] = useReducer(reducer, createInitialState());
  const service = new AppService(dispatch);

  return (
    <AppContext.Provider value={{ state, service }}>
      <header>
        <AppHeader />
      </header>
      <main>
        <MasterDetail />
      </main>
    </AppContext.Provider>
  );
}

export default App;
