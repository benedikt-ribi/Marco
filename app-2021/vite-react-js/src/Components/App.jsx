import { AppContext } from "../Core/AppContext";
import { AppService } from "../../../common/AppService.d.ts";
import { createInitialState, reducer } from "../../../common/appFunctions.d.ts";
import { useReducer } from "react";
import { AppHeader } from "./AppHeader";
import { MasterDetail } from "./MasterDetail";
import "../../../reset.css";
import "../../../styles.css";

function App() {
  const [state, dispatch] = useReducer(reducer, createInitialState());
  const service = new AppService(dispatch);

  return (
    <AppContext.Provider value={{ state, service }}>
      <header>
        <AppHeader />
      </header>
      <main>
        <MasterDetail />
      </main>
    </AppContext.Provider>
  );
}

export default App;
