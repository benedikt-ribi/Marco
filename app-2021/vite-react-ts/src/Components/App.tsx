import { AppContext } from "../Core/AppContext";
import { AppService } from "../Core/AppService.js";
import { createInitialState, reducer } from "../Core/appFunctions.js";
import { useReducer } from "react";
import { AppHeader } from "./AppHeader";
import { MasterDetail } from "./MasterDetail";
import "../../../reset.css";
import "../../../styles.css";

function App() {
  const [state, dispatch] = useReducer(reducer, createInitialState());
  const service = new AppService(dispatch);

  return (
    <AppContext.Provider value={{ state, service }}>
      <header>
        <AppHeader />
      </header>
      <main>
        <MasterDetail />
      </main>
    </AppContext.Provider>
  );
}

export default App;
