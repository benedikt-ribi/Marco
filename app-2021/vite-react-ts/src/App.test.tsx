import { describe, expect, test } from "@jest/globals";
import React from "react";
import { render } from "@testing-library/react";
import App from "./Components/App";

describe("App Component", () => {
  test("App initial contents", () => {
    const { container } = render(<App />);

    expect(container.innerHTML).toMatchSnapshot();
  });
});
