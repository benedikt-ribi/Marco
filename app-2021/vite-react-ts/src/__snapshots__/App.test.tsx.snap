// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component App initial contents 1`] = `"<header><h1>Addresses</h1><ul><li><a href=\\"#\\">Addresses</a></li><li><a href=\\"#\\">Cars</a></li><li><a href=\\"#\\">Albums</a></li><li><a href=\\"#\\">Notes</a></li></ul></header><main><div>loading...</div></main>"`;
